### PR TITLE
executor: fix IndexMerge worker panic because of closed channel

### LINKED ISSUE #29610: panic: send on closed channel
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
panic: send on closed channel [recovered]
        panic: send on closed channel

goroutine 766775274 [running]:
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).handleHandlesFetcherPanic.func1(0x38f2ac0, 0x420f910)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:619 +0x205
github.com/pingcap/tidb/util.WithRecovery.func1(0xc2faf42f30)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/util/misc.go:92 +0x354
panic(0x38f2ac0, 0x420f910)
        /usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/pingcap/tidb/executor.(*partialTableWorker).syncErr(...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:432
github.com/pingcap/tidb/executor.(*partialTableWorker).fetchHandles(0xc02cdabdd0, 0x429f548, 0xc2ec561ec0, 0xc17c86c720, 0xc13bf13320, 0xc13bf13200, 0xc17c86c5a0, 0x42cd6f8, 0xc2eb6c8d20, 0xa, ...)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:448 +0x678
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startPartialTableWorker.func1.1()
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:374 +0x5ba
github.com/pingcap/tidb/util.WithRecovery(0xc3295def08, 0xc2faf42f30)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/util/misc.go:100 +0x4f
github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startPartialTableWorker.func1(0x429f5f0, 0xc2f818d950, 0xc1e8ad8900, 0xc0ff91f0e0, 0x0, 0xc2f480d540, 0x1, 0x1, 0xc17c86c720, 0xc13bf13320)
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:326 +0x1e5
created by github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startPartialTableWorker
        /home/jenkins/agent/workspace/optimization-build-tidb-linux-amd/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:323 +0x238
```
### 4. What is your TiDB version? (Required)

```
Release Version: v5.3.0
Edition: Community
Git Commit Hash: bdc9987370d6d393cb8acae22e238e3350e448b5
Git Branch: heads/refs/tags/v5.3.0
UTC Build Time: 2021-11-03 13:20:53
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #29610

Problem Summary: 
When `partial index worker` got error, it will write give err to `main worker` using `resultCh`. But `resultCh` may already be closed(Maybe because `process worker` finished and close it).

image

### What is changed and how it works?
1. Add a log to indicate why we got error when fetching handles.
2. `partial index worker` wirte its err msg using `fetchCh` instead of using `resultCh` directly.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below): this is covered by QA team.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
executor: fix IndexMerge worker panic because of closed channel
```

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -99,6 +99,7 @@ type lookupTableTask struct {
 	// Step 4   is  completed in "IndexLookUpExecutor.Next".
 	memUsage   int64
 	memTracker *memory.Tracker
+	err        error
```
Comment (wshwsh12) [inline executor/distsql.go:102]: Why not use `doneCh` to send error?

File: executor/index_merge_reader.go [hunk 2]
```diff
@@ -719,6 +721,13 @@ func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, fetchCh <-chan
 
 	distinctHandles := make(map[int64]*kv.HandleMap)
 	for task := range fetchCh {
+		if task.err != nil {
+			// Tell main routine to finish. And continue drain fetchCh.
+			task.doneCh = make(chan error, 1)
+			task.doneCh <- task.err
+			resultCh <- task
+			continue
```
Comment (wshwsh12) [inline executor/index_merge_reader.go:729]: Should we stop processing the remained task? 
Continue or break or something else?

File: executor/distsql.go (hunk 1 above)
Comment (guo-shaoge) [reply executor/distsql.go:102]: make sence.

File: executor/index_merge_reader.go (hunk 2 above)
Comment (guo-shaoge) [reply executor/index_merge_reader.go]: done

File: executor/index_merge_reader.go [hunk 3]
```diff
@@ -789,12 +804,13 @@ type partialIndexWorker struct {
 	partition    table.PhysicalTable // it indicates if this worker is accessing a particular partition table
 }
 
-func (w *partialIndexWorker) syncErr(resultCh chan<- *lookupTableTask, err error) {
-	doneCh := make(chan error, 1)
-	doneCh <- err
-	resultCh <- &lookupTableTask{
-		doneCh: doneCh,
+func (w *partialIndexWorker) syncErr(ctx context.Context, fetchCh chan<- *lookupTableTask, err error) {
+	logutil.Logger(ctx).Error("got error when read index handles", zap.Error(err))
```
Comment (XuHuaiyu) [inline executor/index_merge_reader.go:808]: what's `index handles`?

File: executor/index_merge_reader_test.go [hunk 4]
```diff
@@ -174,6 +175,25 @@ func (s *testSuite1) TestPartitionTableRandomIndexMerge(c *C) {
...
+func (s *testSuite1) TestIndexMergePanicAndError(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t1, t2;")
+	tk.MustExec("create table t1 (c1 int, c2 int, primary key(c1), key(c2));")
+	sql := "select /*+ use_index_merge(t1) */ * from t1 where c1 < 10 or c2 < 10;"
+
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/executor/partialIndexWorkerError", "return"), IsNil)
+	err := tk.QueryToErr(sql)
```
Comment (XuHuaiyu) [inline executor/index_merge_reader_test.go:187]: check the error message here

File: executor/index_merge_reader.go [hunk 5]
```diff
@@ -684,7 +688,7 @@ func (e *IndexMergeReaderExecutor) handleHandlesFetcherPanic(ctx context.Context
 		logutil.Logger(ctx).Error(err4Panic.Error())
 		doneCh := make(chan error, 1)
 		doneCh <- err4Panic
-		resultCh <- &lookupTableTask{
+		ch <- &lookupTableTask{
```
Comment (XuHuaiyu) [inline executor/index_merge_reader.go:691]: Add a failpoint test for this panic

**Review by XuHuaiyu — COMMENTED** [review]
We'd better to add more failpoint test cases

**Comment by guo-shaoge** [comment]
https://github.com/pingcap/tidb/pull/41633
https://github.com/pingcap/tidb/pull/41610
https://github.com/pingcap/tidb/pull/41036

**Pull request closed without merging**